Bugfix when replacing index values leads to duplicates

danielhuppmann opened the pull request:
# Please confirm that this PR has done the following:

- [x] Tests Added
- [x] Documentation Added

# Description of PR

This fixes the bug reported by @Rlamboll in #430. The replacing does not do transitive renaming within one operation, similar to the behaviour of `pandas.Series.replace()`.

closes #430

Rlamboll reviewed (approved):
I guess this is fine, but I wonder how concerned we are with speed here. If we are trying to speed things up, I'd imagine not assigning values to _k and _v for a single use would be more efficient. I also wonder if performing the switch via an if statement rather than try-except would be faster and safer (particularly if no one has ever had success the first way before).  Also, is df["scenario"] = [list] future-proof? I thought we were discouraging that notation.

danielhuppmann commented:
> ... not assigning values to _k and _v for a single use would be more efficient. 

Just tried the following comparison.

```
x = list(range(10**5))

[True if i == x.index(9999) else False for i in range(10**5)]

_i = x.index(9999)
[True if i == _i else False for i in range(10**5)]
```
The second one is orders of magnitude faster, because the expression in the list comprehension isn't executed 10**5 times (but only once).

> I also wonder if performing the switch via an if statement rather than try-except would be faster and safer 

Try-except is faster and I don't see why it wouldn't be unsafe.

> is df["scenario"] = [list] future-proof? I thought we were discouraging that notation.

This is operating on the `DataFrame`, not an `IamDataFrame`.

Rlamboll commented:
OK, I tried it out with implicit _k and _v and it is indeed the same speed or slower, so go ahead and merge.

danielhuppmann commented:
@Rlamboll, I have been thinking about the transitive renaming that would have been allowed by my first implementation. So I refactored it and added a specific test for it (third parametrization of `test_replace_index_level`).

Can you test again in a real-world example if this now works as expected?

Rlamboll commented:
Yes, that behaves as I would expect on my test case. It also looks a lot better now and is easier to follow the flow of logic, rather than needing to work out how valueerrors could occur in the code.

danielhuppmann commented:
thank you for the review!
